src/audio/module_adapter/module/volume/volume_hifi3.c

```diff
@@ -95,46 +95,38 @@ static void vol_s24_to_s24_s32(struct processing_module *mod, struct input_strea
 	bsource->consumed += VOL_S32_SAMPLES_TO_BYTES(samples);
 	bsink->size += VOL_S32_SAMPLES_TO_BYTES(samples);
 
-	while (samples) {
-		m = VOL_BYTES_TO_S32_SAMPLES(audio_stream_bytes_without_wrap(source, in));
-		n = MIN(m, samples);
-		m = VOL_BYTES_TO_S32_SAMPLES(audio_stream_bytes_without_wrap(sink, out));
-		n = MIN(m, n);
-		inu = AE_LA64_PP(in);
-		/* process two continuous sample data once */
-		for (i = 0; i < n; i += 2) {
-			/* Load the volume value */
-			AE_L32X2_XC(volume, vol, inc);
-
-			/* Load the input sample */
-			AE_LA32X2_IP(in_sample, inu, in);
-
-			/* Multiply the input sample */
+	n = samples;
+	inu = AE_LA64_PP(in);
+	/* process two continuous sample data once */
+	for (i = 0; i < n; i += 2) {
+		/* Load the volume value */
+		AE_L32X2_XC(volume, vol, inc);
+
+		/* Load the input sample */
+		AE_LA32X2_IP(in_sample, inu, in);
+
+		/* Multiply the input sample */
 #if COMP_VOLUME_Q8_16
-			out_sample = AE_MULFP32X2RS(AE_SLAI32S(volume, 7), AE_SLAI32(in_sample, 8));
+		out_sample = AE_MULFP32X2RS(AE_SLAI32S(volume, 7), AE_SLAI32(in_sample, 8));
 #elif COMP_VOLUME_Q1_23
-			out_sample = AE_MULFP32X2RS(volume, AE_SLAI32(in_sample, 8));
+		out_sample = AE_MULFP32X2RS(volume, AE_SLAI32(in_sample, 8));
 #else
 #error "Need CONFIG_COMP_VOLUME_Qx_y"
 #endif
 
-			/* Shift for S24_LE */
-			out_sample = AE_SLAI32S(out_sample, 8);
-			out_sample = AE_SRAI32(out_sample, 8);
+		/* Shift for S24_LE */
+		out_sample = AE_SLAI32S(out_sample, 8);
+		out_sample = AE_SRAI32(out_sample, 8);
 
-			/* Store the output sample */
-			AE_SA32X2_IP(out_sample, outu, out);
+		/* Store the output sample */
+		AE_SA32X2_IP(out_sample, outu, out);
 
-			/* calc peak vol
-			 * TODO: fix channel value
-			 */
-			peak_vol_calc(cd, out_sample, 0);
-		}
-		AE_SA64POS_FP(outu, out);
-		samples -= n;
-		in = audio_stream_wrap(source, in);
-		out = audio_stream_wrap(sink, out);
+		/* calc peak vol
+		 * TODO: fix channel value
+		 */
+		peak_vol_calc(cd, out_sample, 0);
```

Comment:
I don't understand all the intrinsics, so I might be wrong, but I don't see where we move from one sample to the next?

Reply:
Thank you all, sorry, I under estimate the difficulty for this change, please give me some time to modify it, it need take time, before that, please hold on your review.

Thanks
Tim